### PR TITLE
feat: Config feature branch

### PR DESCRIPTION
**Pull request by davidterpay**
In this PR, i modifiy our existing configuration set up such that we only need to track two config files
1. one for the app, which determines what server address is expected for the oracle side car and some metrics configurations.
2. one for the oracle side car config. Note that each provider now has a market config that outlines how the on-chain and off-chain currency pairs are represented. This prevents us from having to use a separate file for these configurations.

### CODE REVIEW

**Comment by davidterpay** [comment]
This is going to be a breaking change until all subsequent PRs are merged into this branch.

File: oracle/config/market.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+// CurrencyPairMarketConfig is the config the provider uses to create mappings
+// between on-chain and off-chain currency pairs.
+type CurrencyPairMarketConfig struct {
+	// Ticker is the ticker symbol for the currency pair off-chain.
+	Ticker string `mapstructure:"ticker" toml:"ticker"`
+
+	// CurrencyPair is the on-chain representation of the currency pair.
+	CurrencyPair oracletypes.CurrencyPair `mapstructure:"currency_pair" toml:"currency_pair"`
+}
```
Comment (davidterpay) [inline oracle/config/market.go:36]: We can later extend this to do something like provider specific cross market aggregation, or inversion by market (i.e. we want ETH/BTC but only BTC/ETH is available).

**Comment by davidterpay** [comment]
In a later PR im going to deprecate all of the existing configs and replace it with something like this.

```golang
// LocalConfig defines a readable config for local development. Any changes to this
// file should be reflected in oracle.toml. To update the oracle.toml file, run
// `make update-local-config`. This will update any changes to the oracle.toml file
// as they are made to this file.
var LocalConfig = config.OracleConfig{
	// -----------------------------------------------------------	//
	// --------------------All Currency Pairs---------------------	//
	// -----------------------------------------------------------	//
	CurrencyPairs: []oracletypes.CurrencyPair{
		oracletypes.NewCurrencyPair("BITCOIN", "USD"),
		oracletypes.NewCurrencyPair("ETHEREUM", "USD"),
		oracletypes.NewCurrencyPair("ATOM", "USD"),
		oracletypes.NewCurrencyPair("SOLANA", "USD"),
		oracletypes.NewCurrencyPair("CELESTIA", "USD"),
		oracletypes.NewCurrencyPair("AVAX", "USD"),
		oracletypes.NewCurrencyPair("DYDX", "USD"),
		oracletypes.NewCurrencyPair("ETHEREUM", "BITCOIN"),
		oracletypes.NewCurrencyPair("OSMOSIS", "USD"),
	},
	Production: false,
	// -----------------------------------------------------------	//
	// ----------------------Metrics Config-----------------------	//
	// -----------------------------------------------------------	//
	Metrics: config.MetricsConfig{
		Enabled:                 true,
		PrometheusServerAddress: "localhost:8000",
	},
	UpdateInterval: 1 * time.Second,
	Providers: []config.ProviderConfig{
		{
			// -----------------------------------------------------------	//
			// ---------------------Start API Providers--------------------	//
			// -----------------------------------------------------------	//
			//
			// NOTE: Some of the provider's are only capable of fetching data for a subset of
			// all of the currency pairs. Before adding a new market to the oracle, ensure that
			// the provider supports fetching data for the currency pair.
			Name: binance.Name,
			API: config.APIConfig{
				Name:       binance.Name,
				Atomic:     true,
				Enabled:    true,
				Timeout:    500 * time.Millisecond,
				Interval:   1 * time.Second,
				MaxQueries: 1,
				URL:        binance.US_URL,
			},
			Market: config.MarketConfig{
				Name: binance.Name,
				CurrencyPairToMarketConfigs: map[string]config.CurrencyPairMarketConfig{
					"BITCOIN/USD": {
						Ticker:       "BTCUSDT",
						CurrencyPair: oracletypes.NewCurrencyPair("BITCOIN", "USD"),
					},
					"ETHEREUM/USD": {
						Ticker:       "ETHUSDT",
						CurrencyPair: oracletypes.NewCurrencyPair("ETHEREUM", "USD"),
					},
					"ATOM/USD": {
						Ticker:       "ATOMUSDT",
						CurrencyPair: oracletypes.NewCurrencyPair("ATOM", "USD"),
					},
					"SOLANA/USD": {
						Ticker:       "SOLUSDT",
						CurrencyPair: oracletypes.NewCurrencyPair("SOLANA", "USD"),
					},
					"ETHEREUM/BITCOIN": {
						Ticker:       "ETHBTC",
						CurrencyPair: oracletypes.NewCurrencyPair("ETHEREUM", "BITCOIN"),
					},
				},
			},
		},
		{
			Name: coinbase.Name,
			API: config.APIConfig{
				Name:       coinbase.Name,
				Atomic:     false,
				Enabled:    true,
				Timeout:    500 * time.Millisecond,
				Interval:   1 * time.Second,
				MaxQueries: 5,
				URL:        coinbase.URL,
			},
			Market: config.MarketConfig{
				Name: coinbase.Name,
				CurrencyPairToMarketConfigs: map[string]config.CurrencyPairMarketConfig{
					"BITCOIN/USD": {
						Ticker:       "BTC-USD",
						CurrencyPair: oracletypes.NewCurrencyPair("BITCOIN", "USD"),
					},
					"ETHEREUM/USD": {
						Ticker:       "ETH-USD",
						CurrencyPair: oracletypes.NewCurrencyPair("ETHEREUM", "USD"),
					},
					"ATOM/USD": {
						Ticker:       "ATOM-USD",
						CurrencyPair: oracletypes.NewCurrencyPair("ATOM", "USD"),
					},
					"SOLANA/USD": {
						Ticker:       "SOL-USD",
						CurrencyPair: oracletypes.NewCurrencyPair("SOLANA", "USD"),
					},
					"CELESTIA/USD": {
						Ticker:       "TIA-USD",
						CurrencyPair: oracletypes.NewCurrencyPair("CELESTIA", "USD"),
					},
					"AVAX/USD": {
						Ticker:       "AVAX-USD",
						CurrencyPair: oracletypes.NewCurrencyPair("AVAX", "USD"),
					},
					"DYDX/USD": {
						Ticker:       "DYDX-USD",
						CurrencyPair: oracletypes.NewCurrencyPair("DYDX", "USD"),
					},
					"ETHEREUM/BITCOIN": {
						Ticker:       "ETH-BTC",
						CurrencyPair: oracletypes.NewCurrencyPair("ETHEREUM", "BITCOIN"),
					},
					"OSMOSIS/USD": {
						Ticker:       "OSMO-USD",
						CurrencyPair: oracletypes.NewCurrencyPair("OSMOSIS", "USD"),
					},
				},
			},
		},
		{
			Name: coingecko.Name,
			API: config.APIConfig{
				Name:       coingecko.Name,
				Atomic:     true,
				Enabled:    true,
				Timeout:    500 * time.Millisecond,
				Interval:   15 * time.Second, // Coingecko has a very low rate limit.
				MaxQueries: 1,
				URL:        coingecko.URL,
			},
			Market: config.MarketConfig{
				Name: coingecko.Name,
				CurrencyPairToMarketConfigs: map[string]config.CurrencyPairMarketConfig{
					"BITCOIN/USD": {
						Ticker:       "bitcoin/usd",
						CurrencyPair: oracletypes.NewCurrencyPair("BITCOIN", "USD"),
					},
					"ETHEREUM/USD": {
						Ticker:       "ethereum/usd",
						CurrencyPair: oracletypes.NewCurrencyPair("ETHEREUM", "USD"),
					},
					"ATOM/USD": {
						Ticker:       "cosmos/usd",
						CurrencyPair: oracletypes.NewCurrencyPair("ATOM", "USD"),
					},
					"SOLANA/USD": {
						Ticker:       "solana/usd",
						CurrencyPair: oracletypes.NewCurrencyPair("SOLANA", "USD"),
					},
					"CELESTIA/USD": {
						Ticker:       "celestia/usd",
						CurrencyPair: oracletypes.NewCurrencyPair("CELESTIA", "USD"),
					},
					"DYDX/USD": {
						Ticker:       "dydx-chain/usd",
						CurrencyPair: oracletypes.NewCurrencyPair("DYDX", "USD"),
					},
					"ETHEREUM/BITCOIN": {
						Ticker:       "ethereum/btc",
						CurrencyPair: oracletypes.NewCurrencyPair("ETHEREUM", "BITCOIN"),
					},
					"OSMOSIS/USD": {
						Ticker:       "osmosis/usd",
						CurrencyPair: oracletypes.NewCurrencyPair("OSMOSIS", "USD"),
					},
				},
			},
		},
		// // -----------------------------------------------------------	//
		// // ---------------------Start WebSocket Providers--------------	//
		// // -----------------------------------------------------------	//
		//
		// NOTE: Some of the provider's are only capable of fetching data for a subset of
		// all of the currency pairs. Before adding a new market to the oracle, ensure that
		// the provider supports fetching data for the currency pair.
		{
			Name: cryptodotcom.Name,
			WebSocket: config.WebSocketConfig{
				Name:                cryptodotcom.Name,
				Enabled:             true,
				MaxBufferSize:       1000,
				ReconnectionTimeout: 5 * time.Second,
				WSS:                 cryptodotcom.URL_PROD,
			},
			Market: config.MarketConfig{
				Name: cryptodotcom.Name,
				CurrencyPairToMarketConfigs: map[string]config.CurrencyPairMarketConfig{
					"BITCOIN/USD": {
						Ticker:       "BTCUSD-PERP",
						CurrencyPair: oracletypes.NewCurrencyPair("BITCOIN", "USD"),
					},
					"ETHEREUM/USD": {
						Ticker:       "ETHUSD-PERP",
						CurrencyPair: oracletypes.NewCurrencyPair("ETHEREUM", "USD"),
					},
					"ATOM/USD": {
						Ticker:       "ATOMUSD-PERP",
						CurrencyPair: oracletypes.NewCurrencyPair("ATOM", "USD"),
					},
					"SOLANA/USD": {
						Ticker:       "SOLUSD-PERP",
						CurrencyPair: oracletypes.NewCurrencyPair("SOLANA", "USD"),
					},
					"CELESTIA/USD": {
						Ticker:       "TIAUSD-PERP",
						CurrencyPair: oracletypes.NewCurrencyPair("CELESTIA", "USD"),
					},
					"AVAX/USD": {
						Ticker:       "AVAXUSD-PERP",
						CurrencyPair: oracletypes.NewCurrencyPair("AVAX", "USD"),
					},
					"DYDX/USD": {
						Ticker:       "DYDXUSD-PERP",
						CurrencyPair: oracletypes.NewCurrencyPair("DYDX", "USD"),
					},
					"ETHEREUM/BITCOIN": {
						Ticker:       "ETH_BTC",
						CurrencyPair: oracletypes.NewCurrencyPair("ETHEREUM", "BITCOIN"),
					},
					"OSMOSIS/USD": {
						Ticker:       "OSMO_USD",
						CurrencyPair: oracletypes.NewCurrencyPair("OSMOSIS", "USD"),
					},
				},
			},
		},
		{
			Name: okx.Name,
			WebSocket: config.WebSocketConfig{
				Name:                okx.Name,
				Enabled:             true,
				MaxBufferSize:       1000,
				ReconnectionTimeout: 10 * time.Second,
				WSS:                 okx.URL_PROD,
			},
			Market: config.MarketConfig{
				Name: okx.Name,
				CurrencyPairToMarketConfigs: map[string]config.CurrencyPairMarketConfig{
					"BITCOIN/USD": {
						Ticker:       "BTC-USD",
						CurrencyPair: oracletypes.NewCurrencyPair("BITCOIN", "USD"),
					},
					"ETHEREUM/USD": {
						Ticker:       "ETH-USD",
						CurrencyPair: oracletypes.NewCurrencyPair("ETHEREUM", "USD"),
					},
					"ATOM/USD": {
						Ticker:       "ATOM-USD",
						CurrencyPair: oracletypes.NewCurrencyPair("ATOM", "USD"),
					},
					"SOLANA/USD": {
						Ticker:       "SOL-USD",
						CurrencyPair: oracletypes.NewCurrencyPair("SOLANA", "USD"),
					},
					"CELESTIA/USD": {
						Ticker:       "TIA-USD",
						CurrencyPair: oracletypes.NewCurrencyPair("CELESTIA", "USD"),
					},
					"AVAX/USD": {
						Ticker:       "AVAX-USD",
						CurrencyPair: oracletypes.NewCurrencyPair("AVAX", "USD"),
					},
					"DYDX/USD": {
						Ticker:       "DYDX-USD",
						CurrencyPair: oracletypes.NewCurrencyPair("DYDX", "USD"),
					},
					"ETHEREUM/BITCOIN": {
						Ticker:       "ETH-BTC",
						CurrencyPair: oracletypes.NewCurrencyPair("ETHEREUM", "BITCOIN"),
					},
					"OSMOSIS/USD": {
						Ticker:       "OSMO-USD",
						CurrencyPair: oracletypes.NewCurrencyPair("OSMOSIS", "USD"),
					},
				},
			},
		},
	},
}
```

File: oracle/config/api.go [hunk 2]
```diff
@@ -40,5 +51,13 @@ func (c *APIConfig) ValidateBasic() error {
 		return fmt.Errorf("provider timeout must be greater than 0 and less than the interval")
 	}
 
+	if len(c.URL) == 0 {
```
Comment (aljo242) [inline oracle/config/api.go:54]: https://pkg.go.dev/net/url#ParseRequestURI

Perphaps we also validate the url is valid beyond being empty?

File: oracle/config/app.go [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+
+# ValidatorConsAddress is the validator's consensus address. This is optional. If supplied,
+# the oracle will expose metrics for how often the validator's vote extensions are included
+# in blocks and more.
+validator_cons_address = "{{ .Oracle.ValidatorConsAddress }}"
+`
+)
+
+const (
+	flatEnabled                 = "oracle.enabled"
```
Comment (aljo242) [inline oracle/config/app.go]:
```suggestion
	flagEnabled                 = "oracle.enabled"
```

File: oracle/config/app.go [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+
+// ReadConfigFromAppOpts reads the config parameters from the AppOptions and returns the config.
+func ReadConfigFromAppOpts(opts servertypes.AppOptions) (AppConfig, error) {
+	var (
+		cfg AppConfig
+		err error
+	)
+
+	// determine if the oracle is enabled
+	if v := opts.Get(flatEnabled); v != nil {
```
Comment (aljo242) [inline oracle/config/app.go]:
```suggestion
	if v := opts.Get(flagEnabled); v != nil {
```

File: oracle/config/market.go (hunk 1 above)
Comment (aljo242) [inline oracle/config/market.go:36]: Why can we not just do a map of Ticker -> CurrencyPair ?

File: oracle/config/metrics.go [hunk 5]
```diff
@@ -2,99 +2,29 @@ package config
...
-	if err := viper.Unmarshal(&config); err != nil {
-		return MetricsConfig{}, err
+func (c *MetricsConfig) ValidateBasic() error {
+	if !c.Enabled {
+		return nil
 	}
 
-	if err := config.ValidateBasic(); err != nil {
-		return MetricsConfig{}, err
+	if len(c.PrometheusServerAddress) == 0 {
```
Comment (aljo242) [inline oracle/config/metrics.go]: Can we validate this beyond being empty as well?

File: oracle/config/market.go (hunk 1 above)
Comment (davidterpay) [reply oracle/config/market.go:36]: toml has some strange behavior with the keys of maps (it always lower cases them), so you might be losing some amount of information when you use the ticker as the key. IMO, this also provides better standardization and if you take a look at what the dydx oracle config looks like you'll see that it is very similar.

File: oracle/config/metrics.go (hunk 5 above)
Comment (davidterpay) [reply oracle/config/metrics.go]: added the same parseURL functionality you recommended above.

File: oracle/config/market.go (hunk 1 above)
Comment (davidterpay) [reply oracle/config/market.go:36]: turns out it was an issue on my end, gonna update this.

**Comment by davidterpay** [comment]
https://github.com/spf13/viper/issues/373

un-resolved issue where keys in toml maps always get converted to lower-case. in that case we dont want to use tickers as keys since you might lose some amount of information.

**Review by aljo242 — APPROVED** [review]

**Pull request merged**